source/multigrid/mg_transfer_prebuilt.cc

```diff
@@ -279,7 +280,7 @@ void MGTransferPrebuilt<VectorType>::build_matrices
                                                    mg_dof);
       dsp.reinit(0,0);
 
-      FullMatrix<double> prolongation;
+      FullMatrix<typename VectorType::value_type> prolongation;
```

Comment:
This is the error message:
```
../source/multigrid/mg_transfer_prebuilt.cc:317:19: error: no matching function for call to ‘dealii::PETScWrappers::MPI::SparseMatrix::set(__gnu_cxx::__alloc_traits<std::allocator<long long unsigned int>, long long unsigned int>::value_type&, const unsigned int&, long long unsigned int*, dealii::AlignedVector<double>::value_type*, bool)’
                   prolongation_matrices[level]->set (dof_indices_child[i],
                   ^~~~~~~~~~~~~~~~~~~~~
In file included from ../include/deal.II/lac/petsc_parallel_sparse_matrix.h:24,
                 from ../include/deal.II/multigrid/mg_transfer.h:28,
                 from ../source/multigrid/mg_transfer_prebuilt.cc:37:
../include/deal.II/lac/petsc_matrix_base.h:1202:3: note: candidate: ‘void dealii::PETScWrappers::MatrixBase::set(dealii::PETScWrappers::MatrixBase::size_type, dealii::PETScWrappers::MatrixBase::size_type, PetscScalar)’
   MatrixBase::set (const size_type   i,
   ^~~~~~~~~~
../include/deal.II/lac/petsc_matrix_base.h:1202:3: note:   candidate expects 3 arguments, 5 provided
../include/deal.II/lac/petsc_matrix_base.h:1215:3: note: candidate: ‘void dealii::PETScWrappers::MatrixBase::set(const std::vector<long long unsigned int>&, const dealii::FullMatrix<std::complex<double> >&, bool)’
   MatrixBase::set (const std::vector<size_type>  &indices,
   ^~~~~~~~~~
../include/deal.II/lac/petsc_matrix_base.h:1215:3: note:   candidate expects 3 arguments, 5 provided
../include/deal.II/lac/petsc_matrix_base.h:1232:3: note: candidate: ‘void dealii::PETScWrappers::MatrixBase::set(const std::vector<long long unsigned int>&, const std::vector<long long unsigned int>&, const dealii::FullMatrix<std::complex<double> >&, bool)’
   MatrixBase::set (const std::vector<size_type>  &row_indices,
   ^~~~~~~~~~
../include/deal.II/lac/petsc_matrix_base.h:1232:3: note:   candidate expects 4 arguments, 5 provided
../include/deal.II/lac/petsc_matrix_base.h:1251:3: note: candidate: ‘void dealii::PETScWrappers::MatrixBase::set(dealii::PETScWrappers::MatrixBase::size_type, const std::vector<long long unsigned int>&, const std::vector<std::complex<double> >&, bool)’
   MatrixBase::set (const size_type                 row,
   ^~~~~~~~~~
../include/deal.II/lac/petsc_matrix_base.h:1251:3: note:   candidate expects 4 arguments, 5 provided
../include/deal.II/lac/petsc_matrix_base.h:1267:3: note: candidate: ‘void dealii::PETScWrappers::MatrixBase::set(dealii::PETScWrappers::MatrixBase::size_type, dealii::PETScWrappers::MatrixBase::size_type, const size_type*, const PetscScalar*, bool)’
   MatrixBase::set (const size_type    row,
   ^~~~~~~~~~
../include/deal.II/lac/petsc_matrix_base.h:1267:3: note:   no known conversion for argument 4 from ‘dealii::AlignedVector<double>::value_type*’ {aka ‘double*’} to ‘const PetscScalar*’ {aka ‘const std::complex<double>*’}
../source/multigrid/mg_transfer_prebuilt.cc: In instantiation of ‘void dealii::MGTransferPrebuilt<VectorType>::build_matrices(const dealii::DoFHandler<dh_dim, spacedim>&) [with int dim = 3; int spacedim = 3; VectorType = dealii::PETScWrappers::MPI::Vector]’:
source/multigrid/mg_transfer_prebuilt.inst:138:31:   required from here
../source/multigrid/mg_transfer_prebuilt.cc:317:19: error: no matching function for call to ‘dealii::PETScWrappers::MPI::SparseMatrix::set(__gnu_cxx::__alloc_traits<std::allocator<long long unsigned int>, long long unsigned int>::value_type&, const unsigned int&, long long unsigned int*, dealii::AlignedVector<double>::value_type*, bool)’
                   prolongation_matrices[level]->set (dof_indices_child[i],
```
The problem here is that `std::vector<std::shared_ptr<typename internal::MatrixSelector<VectorType>::Matrix> > MGTransferPrebuilt< VectorType >::prolongation_matrices` is has `PetscScalar` entries and the [correct overload](https://dealii.org/developer/doxygen/deal.II/classPETScWrappers_1_1MatrixBase.html#ae9fe7cc7dd404261b63b7ee651233de1) for `set` expects a `PetscScalar` pointer.

Reply:
I see. So the assignment I pointed out then converts the real matrix to a complex-valued matrix?

I think the original problem may have been to make the `prolongation_matrices` of type `PetscScalar`. I'd think that prolongation matrices should always be real, though they can be applied to complex-valued vectors of course. Do you know why the `prolongation_matrices` have this type?

Reply:
> Do you know why the prolongation_matrices have this type?

They need to be compatible with `VectorType` such that `vmult` is implemented.

Reply:
> They need to be compatible with VectorType such that vmult is implemented.

One way is to add `vmult` between real-valued matrix and a complex-valued vector.

Reply:
Is this going to work when you have are dealing with a `PETSc` matrix and a `PETSc` vector?  Can you use different scalar types?

Reply:
> Is this going to work when you have are dealing with a PETSc matrix and a PETSc vector? Can you use different scalar types?

ah, we are speaking about global (instead of cell-based) prolongation matrices, i.e. level PETSc vector -> level PETSc vector. No, mixing real and complex with PETSc won't work, if it's compiled in complex algebra, everything is complex valued.

Reply:
Yes, that's a shame. I think it's entirely unnecessary here to use complex-valued data types, but I also don't see a way around it. So maybe leave a comment about why we don't use a `FullMatrix<double>` above the declaration of the variable, summarizing the discussion here?